pkg/kubelet/network/kubenet/kubenet_linux.go

```diff
@@ -456,3 +484,317 @@ func (plugin *kubenetNetworkPlugin) getNsenterPath() (string, error) {
+	}
+	args := []string{"-m", "comment", "--comment", "kubenet hostport portals",
+		"-m", "addrtype", "--dst-type", "LOCAL",
+		"-j", string(kubenetHostportChain)}
+	for _, tc := range tableChainsNeedJumpServices {
+		if _, err := plugin.iptables.EnsureRule(utiliptables.Prepend, tc.table, tc.chain, args...); err != nil {
+			glog.Errorf("Failed to ensure that %s chain %s jumps to %s: %v", tc.table, tc.chain, kubenetHostportChain, err)
+			return
+		}
+	}
+	// Need to SNAT traffic from localhost
+	args = []string{"-m", "comment", "--comment", "kubenet postrouting rules", "-o", BridgeName, "-s", "127.0.0.1/8", "-j", "MASQUERADE"}
+	if _, err := plugin.iptables.EnsureRule(utiliptables.Prepend, utiliptables.TableNAT, utiliptables.ChainPostrouting, args...); err != nil {
+		glog.Errorf("Failed to ensure that %s chain %s jumps to MASQUERADE: %v", utiliptables.TableNAT, utiliptables.ChainPostrouting, err)
+		return
+	}
+
+	// Get iptables-save output so we can check for existing chains and rules.
+	// This will be a map of chain name to chain with rules as stored in iptables-save/iptables-restore
+	existingNATChains := make(map[utiliptables.Chain]string)
+	iptablesSaveRaw, err := plugin.iptables.Save(utiliptables.TableNAT)
+	if err != nil { // if we failed to get any rules
+		glog.Errorf("Failed to execute iptables-save, syncing all rules: %v", err)
+	} else { // otherwise parse the output
+		existingNATChains = utiliptables.GetChainLines(utiliptables.TableNAT, iptablesSaveRaw)
+	}
+
+	natChains := bytes.NewBuffer(nil)
+	natRules := bytes.NewBuffer(nil)
+	writeLine(natChains, "*nat")
+	// Make sure we keep stats for the top-level chains, if they existed
+	// (which most should have because we created them above).
+	if chain, ok := existingNATChains[kubenetHostportChain]; ok {
+		writeLine(natChains, chain)
+	} else {
+		writeLine(natChains, utiliptables.MakeChainLine(kubenetHostportChain))
+	}
+
+	// Accumulate NAT chains to keep.
+	activeNATChains := map[utiliptables.Chain]bool{} // use a map as a set
+
+	for containerPort, target := range containerPortMap {
+		protocol := strings.ToLower(string(containerPort.Protocol))
+		hostportChain := hostportChainName(containerPort, target.podFullName)
+		if chain, ok := existingNATChains[hostportChain]; ok {
+			writeLine(natChains, chain)
+		} else {
+			writeLine(natChains, utiliptables.MakeChainLine(hostportChain))
+		}
+
+		activeNATChains[hostportChain] = true
+
+		// Capture the hostport
+		args := []string{
+			"-A", string(kubenetHostportChain),
+			"-m", "comment", "--comment", fmt.Sprintf(`"%s hostport %d"`, target.podFullName, containerPort.HostPort),
+			"-m", protocol, "-p", protocol,
+			"--dport", fmt.Sprintf("%d", containerPort.HostPort),
+		}
+		writeLine(natRules, append(args, "-j", string(hostportChain))...)
+
+		// Create hostport chain to DNAT traffic to final destination
+		// Iptables will maintained the stats for this chain
+		args = []string{
+			"-A", string(hostportChain),
```

Comment:
you end up jumping to a 1-rule chain which is noisier and less efficient.  Can we just flatten this rule into the KUBE-HOSTPORTS chain so it does the dport check and then jumps to DNAT directly?